stl/inc/ranges

```diff
@@ -620,6 +620,132 @@ namespace ranges {
+    template <_Objectable _Ty>
+    class _Non_propagating_cache { // a simplified optional that resets on copy / move
+        // clang-format on
+    public:
+        constexpr _Non_propagating_cache() noexcept : _Dummy{} {}
+
+        _CONSTEXPR20_DYNALLOC ~_Non_propagating_cache() {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Dummy = {};
+            }
+        }
+
+        constexpr _Non_propagating_cache(const _Non_propagating_cache&) noexcept {}
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache(_Non_propagating_cache&& _Other) noexcept {
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Dummy   = {};
+                _Other._Engaged = false;
+            }
+        }
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache& operator=(const _Non_propagating_cache& _Other) noexcept {
+            if (_STD addressof(_Other) == this) {
+                return *this;
+            }
+
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Dummy   = {};
+                _Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache& operator=(_Non_propagating_cache&& _Other) noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Dummy   = {};
+                _Engaged = false;
+            }
+
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Dummy   = {};
+                _Other._Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+
+        template <class _It>
+        _CONSTEXPR20_DYNALLOC _Ty& _Emplace_deref(_It& _Iter) noexcept(noexcept(_Ty(*_Iter))) /* strengthened */ {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Dummy   = {};
+                _Engaged = false;
+            }
+
+            _Construct_in_place(_Val, *_Iter);
+            _Engaged = true;
+
+            return _Val;
+        }
+
+    private:
+        union {
+            _Nontrivial_dummy_type _Dummy;
+            _Ty _Val;
+        };
+        bool _Engaged = false;
+    };
+
+    // clang-format off
+    template <_Objectable _Ty>
+        requires is_trivially_destructible_v<_Ty>
+    class _Non_propagating_cache<_Ty> { // provide the same API more efficiently for trivially destructible types
+        // clang-format on
+    public:
+        constexpr _Non_propagating_cache() noexcept {}
+
+        constexpr _Non_propagating_cache(const _Non_propagating_cache&) noexcept {}
+
+        constexpr _Non_propagating_cache(_Non_propagating_cache&& _Other) noexcept {}

```

Comment:
Tests are failing because we've named unused parameters here, on 724, and on 728.